Feat/four oh four 2

bean00 opened the pull request:
I'm still not completely satisfied with how my code is structured and organized right now. In particular, I plan on building out the Request class to be returned from `readRequest` in the `MessageController`.

The `README` will be done by tomorrow (Tuesday 1/22).

Notes on the design:
- `RequestReader` reads in the request from the input stream
- `RequestParser` parses the request
- `RequestInterpreter` interprets the request to determine the appropriate status code to return

Commit:
Get Simple Get test to pass

Commit:
Get Four Oh Four test to pass

jdesrosiers commented on src/main/java/com/bean00/MessageController.java:
```diff
@@ -1,32 +1,58 @@
+        String simple200Response = response.simple200Response();
+        String simple404Response = response.simple404Response();
+
+        List<String> responses = List.of(simple200Response, simple404Response);
```
Is this list going to grow to encompass all possible responses?

bean00 replied:
No, this wouldn't be a good idea. The `Response` class would need to be updated to generate the strings using the different elements of a response.

jdesrosiers commented on src/main/java/com/bean00/MessageController.java:
```diff
@@ -1,32 +1,58 @@
+        String statusCode = Integer.toString(statusCodeInt);
+        List<String> responses = createListOfResponses();
+
+        String desiredResponse = responses
```
It doesn't look like a list is the right choice of data structure for `responses`. What is a better choice? Why?

bean00 replied:
I think it would actually be better to modify the `Response` object so that it contains the necessary data to generate a string representation of the response.

tbaik commented on src/main/java/com/bean00/RequestReader.java:
```diff
@@ -0,0 +1,39 @@
+
+    public ArrayList<String> getHeaders() throws IOException {
+        ArrayList<String> headers = new ArrayList<>();
+        String header;
```
I think it's better to declare this on line 23 inside the while loop if you don't need it outside of the while loop!

bean00 replied:
So I tried doing that (change it to `String header = in.readLine();` as the first statement inside the while loop), but there was an error (`Error: java: cannot find symbol symbol: variable header`). I believe the `header` variable still needs to be declared outside of the loop, because the check for the while is outside the loop.

tbaik replied:
ohh woops, I did not see that it's being used there. My bad. I think you can put line 23 inside the while loop too actually, but upto you!

tbaik commented on src/main/java/com/bean00/Response.java:
```diff
@@ -1,9 +1,34 @@
+import java.util.HashMap;
+
 public class Response {
+    private String version = "HTTP/1.1";
```
can you try making this immutable?

bean00 replied:
I made it a constant using `private static final`.

tbaik reviewed (changes requested):
I think it looks really good overall! I like the abstractions that are forming and I'm sure they'll evolve as you build this out. I just had a few small things for you to consider.

bean00 commented:
So I've addressed the comments for this PR. One thing that came up is that the Cob Spec tests weren't passing. I believe it has something to do with the `PrintWriter` `print` method. When I used `println` instead and took off the trailing `\n` from the Response string, then the Cob Spec tests worked. I'll keep it like that for now, and will work on moving over to use `ByteArrayInputStream` and `ByteArrayOutputStream` instead.

jdesrosiers commented:
I'm approving changes, but make sure cobspec is passing what it should be passing before merging.
